[ENH] Lagged window summarizer transformation

danbartl opened the pull request:
Reference Issues/PRs
Partitoned implementation of one part of global forecasting  #1870

What does this implement/fix? Explain your changes.
Added a new Transformer that applies a transformation of a target series by applying rolling window functions, e.g. a rolling mean across a window of 7 shifted by 7. This will greatly boost performance of tree based models applied on time series data.

What should a reviewer concentrate their feedback on?
Key of the PR is the provision of functions to be applied across certain past windows of a time series. These functions can be defined in a dictionary giving

- name of the feature
- function to be used (will default to native pandas implementations, if available
- window shift (shift 0 refers to first observation in past)
- window length


E.g.:
```
kwargs = {
    "lag_config": {
        "mean": ["mean", [[7, 0], [14, 0]]],
        "covar_feature": ["cov", [[28, 0]],
    }
}
```

Would result in features corresponding to 

- column "mean_7_0", the mean of the last 7 days
- column "mean_7_7", the mean of the 7 days prior to the last 7 days
- column "covar_feature_28_0", the covariance of the last 28 days

Does the format for the dictionary make sense? 

For new estimators
I've added the estimator to the online documentation.
not yet:  I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

fkiraly commented:
updated `convert_to` to behave properly if you have multiple internal mtypes that can be of different scitype, in https://github.com/alan-turing-institute/sktime/pull/1929 - let me know if that works/helps

ngupta23 commented:
@danbartl Tring to understand this better...

`"mean": ["mean", [8, 7]]`

Does this mean that the window length = 7 starting at lag = 8? So if each `-` or `x` represents a point in the time series, then does this take the mean of the points marked in `x`?

\- - - - - - - - - - - - x x x x x x x - - - - - - - 

If my understanding is correct, should this not be lag=7 instead of lag=8 since the last point would be lag=0?

Thanks.

fkiraly commented on sktime/transformations/series/tests/test_window_summarizer.py:
```diff
@@ -0,0 +1,122 @@
+transformer_empty = LaggedWindowSummarizer(**kwargs_empty)
+
+# Check if transformers works with non-Panel date
+Xt_univar = transformer.fit_transform(y_train)
```
substantial functionality like this should happen *inside* a test function, or this may error out in test collection (not in execution)

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,245 @@
+        super(_LaggedWindowExtractor, self).__init__()
+
+    # Get extraction parameters
+    def fit(self, X, y=None):
```
this should be `_fit`, no?

danbartl replied:
done

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,245 @@
+
+    # Get extraction parameters
+    def fit(self, X, y=None):
+        """Fit.
```
copy the docstring from other `_fit` (see extension template)

fkiraly reviewed (changes requested):
Looks great, and nice idea!

I left some comments on the tests.
You should also use `_fit` in the estimator.

fkiraly commented on sktime/utils/validation/series.py:
```diff
@@ -18,7 +18,13 @@
+    pd.RangeIndex,
+    pd.PeriodIndex,
+    pd.DatetimeIndex,
+    pd.MultiIndex,
```
I don't think we want to change that in this PR

fkiraly replied:
(should go in the hierarchical PR)

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,245 @@
+            * "skew",
+            * "sem"
+         or function definition
+    window: list of integers
```
you need to write more detail here, or it will be hard to understand what it does.
Use math notation or array notation.

fkiraly commented:
> @danbartl Trying to understand this better...

@ngupta23, I think so, according to an earlier design.
One of my comment points is the docstring, this needs better explaining.
@danbartl may appreciate your feedback here, e.g., whether it's easy to understand or not what the parameters are doing.

ngupta23 commented:
> @danbartl may appreciate your feedback here, e.g., whether it's easy to understand or not what the parameters are doing.

@fkiraly Yes, we discussed offline on Slack (sorry for multiple threads). He agreed with my intuition and is going to have a look at it.

Honestly, the current nomenclature was a little confusing, hence my question. It might also be good to align this as a standard convention (pseudocode) of [window_size, starting_at=lags] as more people will be familiar with that.

Also, (although unrelated) that will make the inputs consistent with something like what is entered for `Differencer`, where the user has to enter the lags.

fkiraly commented:
ah
well, that's why I wanted to see the docstring, as to what was *intended*

fkiraly commented:
CI/CD issues are now resolved - updated from main and tests should now run.
Let us know when ready for review.

fkiraly commented:
@ltsaprounis, the test failure is caused by the docstring example - this should be a self-contained piece of code that executes without exception. It may also be a whitespace/indentation issue, kindly check.

ltsaprounis commented:
examples should work now.
@danbartl if you want to test them locally and quickly try this: 
`python -m doctest -o=ELLIPSIS -v sktime/transformations/panel/window_summarizer.py`

Also, I think doctest doesn't like `Parallel` so I added `n_jobs = 1`

fkiraly commented:
oh, apologies - I mis-tagged, I wanted to tag @danbartl. Not sure why, perhaps "Lagged TRansformation" is similar to @LTsapRounis and my brain melted.

Anyway, thanks for fixing!

fkiraly commented on sktime/transformations/panel/window_summarizer.py:
```diff
@@ -0,0 +1,268 @@
+    """
+
+    _tags = {
+        # todo: what is the scitype of X: Series, or Panel
```
remove the "todo" from where you've filled in the right thing

fkiraly commented on sktime/transformations/panel/window_summarizer.py:
```diff
@@ -0,0 +1,268 @@
+        Private _fit containing the core logic, called from fit
+        Parameters
+        ----------
+        X : Series or Panel
```
change this to the type expected internally

danbartl replied:
Not sure, the transform can take Series and Pandas DataFrame as inputs and will return a Pandas Dataframe (typically one Series / column gets transformed to multiple columns). Panel is not supported due to the changes required that I factored out (part of hierarchical PR). So what should I choose here?

fkiraly replied:
this is the docstring of the inner function, so it should say
`X : pandas.DataFrame or pandas.Series`
and `pandas.DataFrame` for the return object.

fkiraly commented on sktime/transformations/panel/window_summarizer.py:
```diff
@@ -0,0 +1,268 @@
+        # todo: what is the scitype of X: Series, or Panel
+        # "scitype:transform-input": "Panel",
+        # todo: what scitype is returned: Primitives, Series, Panel
+        "scitype:transform-output": "Panel",
```
should this not be `Series`? Since we transform the series into a series that has the lagged summaries?

danbartl replied:
Not sure, the transform can take Series and Pandas DataFrame as inputs and will return a Pandas Dataframe (typically one Series / column gets transformed to multiple columns). Panel is not supported due to the changes required that I factored out (part of hierarchical PR). So what should I choose here?

fkiraly replied:
I think it should be `Series`, that´s for all univariate series formats (`pd.Series` and `pd.DataFrame`) - let´s see what happens...

danbartl replied:
ok just tried:

        "scitype:transform-input": "Series",
        "scitype:transform-output": "Series",

This works in principle , but then the following happens:

-  Input pd.DataFrame with one argument in kwargs -> pd.DataFrame Output with one column
- Input pd.DataFrame with multiple arguments in kwargs -> pd.DataFrame Output with multiple column
- Input pd.Series with one argument in kwargs -> **pd.Series Output** with one column
- Input pd.Series with multiple arguments in kwargs -> pd.DataFrame Output with multiple column


I personally would prefer to have always the same output (pd.DataFrame), especially since it now depends on kwargs argument. How can I make sure this happens for case 3? Currently the output gets automatically converted.

Thanks!

fkiraly replied:
> I personally would prefer to have always the same output (pd.DataFrame), especially since it now depends on kwargs argument. What do you think?

Currently, what you describe is *default behaviour* of the interface, of all forecasters.

It has been discussed internally and decided that this is a bit confusing, and the output should always be `pd.DataFrame`. The change will likely happen in 0.11.0, through changes in the `BaseForecaster`. I.e., it will automatically be applied to all forecasters and yours through the `BaseForecaster` change.

fkiraly replied:
current behaviour is "input type = output type (as long as it is possible)", it will be replaced with "constant output type"

fkiraly commented on sktime/transformations/panel/window_summarizer.py:
```diff
@@ -0,0 +1,268 @@
+    return feat
+
+
+class _LaggedWindowExtractor(_PanelToTabularTransformer):
```
why are there two classes? Seems one adds `_fit`, the oder one `_transform`, so why is not both in one single class?

fkiraly commented on sktime/transformations/panel/window_summarizer.py:
```diff
@@ -0,0 +1,268 @@
+        else:
+            self._func_dict = None
+            self._truncate_start = None
+        self._is_fitted = True
```
this line is unnecessary, it's done by `fit`

fkiraly reviewed (changes requested):
I left some more comments, thanks for the changes.
Some are just questions for clarification.

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -143,17 +143,9 @@ class LaggedWindowSummarizer(BaseTransformer):
-        # "y_inner_mtype": "None",  # which mtypes do _fit/_predict support for y?
-        "capability:inverse_transform": False,  # does transformer
-        # have inverse transform
+        "scitype:transform-output": "Panel",
```
why that? I thought `Series` was right, can you explain?

danbartl replied:
yes that is correct, but then tests fail currently due to this self._convert_X issue. But probably not the right way to fix I agree. will change it back, just wanted to see if everything else works.

fkiraly replied:
I was just working on it, give me a sec

fkiraly replied:
FYI, here: https://github.com/alan-turing-institute/sktime/pull/1977

Thanks for creating the very specific condition that caused the failure 😄

danbartl replied:
works now

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,357 @@
+        transformed version of X
+        """
+        # input checks
+        if not isinstance(y, pd.DataFrame):
+            y = pd.DataFrame(y)
+
+        if not isinstance(X, pd.DataFrame):
+            X = pd.DataFrame(X)
```
I am wondering about this type casting, should this not be covered by changing this given tags:
```
        "X_inner_mtype": [
            "pd.Series",
            "pd.DataFrame",
        ],  # which mtypes do _fit/_predict support for X?
        "y_inner_mtype": [
            "pd.Series",
            "pd.DataFrame",
        ], 
```
to just `pd.DataFrame`?
I guess @fkiraly should be able to answer this as well?

danbartl replied:
seems to work, thanks for the hint

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,357 @@
+    def _fit(self, X, y=None):
+        """Fit transformer to X and y.
+
+        Private _fit containing the core logic, called from fit
+        Parameters
```
```suggestion
        Private _fit containing the core logic, called from fit

        Parameters
```

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,357 @@
+        Parameters
+        ----------
+        X : Series or Panel
+        y : Series or Panel
+        Returns
```
```suggestion
        y : Series or Panel

        Returns
```

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,357 @@
+    return feat
+
+
+class LaggedWindowSummarizer(BaseTransformer):
```
can we move this class up and the functions below? just a detail, but its more readable then

aiwalter reviewed (commented):
left few minor comments @danbartl

danbartl commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,357 @@
+                    + " was renamed to 'y'."
+                )
+
+        if self.target_cols is not None:
+            if not all(x in Z_name for x in self.target_cols):
+                raise ValueError("target_X targets non-existing column in X")
```
makes sense, changed

danbartl commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,357 @@
+    ...  }
+    >>> transformer = LaggedWindowSummarizer(**kwargs_variant)
+    >>> y_transformed = transformer.fit_transform(y)
+    """
```
the use within make_reduction does not yet work, since the hierarchical PR is not finished. However I included two instructive ForecastingPipeline examples

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,385 @@
+            or function definition
+    window: list of integers
+        Contains values for window shift and window length.
+    target_cols: list of str,
+        Specifies which columns in y or X to target. If set to None will
+        target first column in X.
```
I was wondering about this param what it is actually doing? As `y` is always the target and `X` the exog, why do we need this `taget_cols` param? @danbartl

danbartl replied:
@aiwalter you can target columns to transform in X only, target_cols specifies which columns to transform. If you want to transform y, append it to X and add it to target_cols. target_cols will drop the column, but keep the transforms (like lags etc.) Will change the comment to update that.

danbartl commented:
> left few minor comments @danbartl

Thanks 👍 I slightly reworked the functionalities due to your comments. @fkiraly I addressed all points so far, do you have any additional?

fkiraly commented:
> @fkiraly I addressed all points so far, do you have any additional?

#1977 needs to go in first - @aiwalter, are you happy?

fkiraly commented on sktime/transformations/series/tests/test_window_summarizer.py:
```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+"""Test extraction of features across (shifted) windows."""
+__author__ = ["Daniel Bartling"]
```
should be GitHub ID

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,360 @@
+
+    Examples
+    --------
+    >>> from sktime.transformations.series.window_summarizer import WindowSummarizer
```
could you make sure that the example code is properly formatted?
With all linting rules applied.

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,360 @@
+        "scitype:instancewise": True,
+        "capability:inverse_transform": False,
+        "scitype:transform-labels": "Series",
+        "X_inner_mtype": [
```
move to one line

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,360 @@
+        if self._X_rename is not None:
+            X.columns = [self._X_rename]
+
+        Z = X
```
let's not do this, let's call everything `X`

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,360 @@
+        X.columns = X.columns.map(str)
+
+        if self._X_rename is not None:
+            X.columns = [self._X_rename]
```
this is a side effect on `X`? Let's not rename the columns of `X`.

danbartl replied:
Actually this is a left over from a previous approach with no impact. Removed it.

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,360 @@
+        -------
+        transformed version of X
+        """
+        # input checks
```
remove comments

danbartl replied:
done

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,360 @@
+            else:
+                # for _index, kwargs in self._func_dict.iterrows():
+                #     _window_feature(X, **kwargs)
+                df = Parallel(n_jobs=self.n_jobs)(
```
do we need to make a case distinction for `n_jobs=-1`, or is this fine?

danbartl replied:
Could you explain what you mean by that? n_jobs = -1 is the default and works fine

fkiraly commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,360 @@
+        "scitype:transform-output": "Series",
+        "scitype:instancewise": True,
+        "capability:inverse_transform": False,
+        "scitype:transform-labels": "Series",
```
`y` is not used, so should be `None`

fkiraly reviewed (changes requested):
I left some comments above.
Really nice work, very useful transformer!

There is one blocker:
* I think it is possible that `_transform` has side effects on `X`, through column assignment. There should not be any side effects on arguments. Also, I'm not sure why the tests did not pick this up, we test for side effects - any ideas?

The other things are style/format comments.

danbartl commented:
Actually the renaming was removed in a prior version, sorry about the confusion (it renamed to the already existing name only, but this is removed in the latest version)

danbartl commented:
I tried to address all your comments. The n_jobs = -1 question is uncritical in my view, but maybe you could further clarify. Apart from that everything is finished imo. Thanks for all the comments!

aiwalter commented:
@danbartl could you check code coverage? Its "only" 89%, but I think a transformer could easily have 100%.

danbartl commented:
> @danbartl could you check code coverage? Its "only" 89%, but I think a transformer could easily have 100%.

I think this is mainly coming from the verious pd.MultiIndex checks. I implemented them because WindowSummarizer is crucial for hierarchical use case, but unfortunately that cannot be tested until the hierarchical PRs are concluded. I still would like to keep that in to reference WindowSummarizer  in the PR I am working on in parallel.

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,334 @@
+
+    Parameters
+    ----------
+    Dictionary with the following arguments:
+    index : str, base string of the derived features
+            final name will also contain
+            window shift and window length.
+    summarizer: either str corresponding to implemented pandas function, currently
+            * "sum",
+            * "mean",
+            * "median",
+            * "std",
+            * "var",
+            * "kurt",
+            * "min",
+            * "max",
+            * "corr",
+            * "cov",
+            * "skew",
+            * "sem"
+            or function definition
+    window: list of integers
+        Contains values for window shift and window length.
+    target_cols: list of str,
+        Specifies which columns in X to target. If set to None will
+        target first column in X.
```
I am confused, you say a dict, but we have the params
```
        summarizer=None,
        n_jobs=-1,
        target_cols=None,
```
So its not a dict actually? What you are doing in examples is just gving those 3 params as a kwargs dict.
```suggestion
    index : str, base string of the derived features
            final name will also contain
            window shift and window length.
    summarizer: either str corresponding to implemented pandas function, currently
            * "sum",
            * "mean",
            * "median",
            * "std",
            * "var",
            * "kurt",
            * "min",
            * "max",
            * "corr",
            * "cov",
            * "skew",
            * "sem"
            or function definition
    window: list of integers
        Contains values for window shift and window length.
    target_cols: list of str,
        Specifies which columns in X to target. If set to None will
        target first column in X.
```

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,334 @@
+    >>> from sktime.forecasting.compose import ForecastingPipeline
+    >>> from sktime.forecasting.model_selection import temporal_train_test_split
+    >>> y = load_airline()
+    >>> kwargs = WindowSummarizer.get_test_params()[0]
```
I think this `get_test_params` should not be used in an example? Its not obvious and readable what you hand over here

fkiraly replied:
yes - better write them out explicitly. Good spot!
The `get_test_params` function is only for use in testing, and not in general a good user parameter choice.

aiwalter replied:
maybe we should make it a private method? @fkiraly

aiwalter replied:
I think we should actually, its not meant to be exposed to user

fkiraly replied:
well, the user could use it for testing?

aiwalter commented on sktime/transformations/series/window_summarizer.py:
```diff
@@ -0,0 +1,334 @@
+    >>> kwargs = WindowSummarizer.get_test_params()[0]
+    >>> transformer = WindowSummarizer(**kwargs)
+    >>> y_transformed = transformer.fit_transform(y)
+    >>> # Example y univariate
```
you can split the example into separate sections, would look like this: https://www.sktime.org/en/latest/api_reference/auto_generated/sktime.forecasting.model_selection.ForecastingGridSearchCV.html?highlight=forecastinggridsearch#sktime.forecasting.model_selection.ForecastingGridSearchCV

fkiraly reviewed (changes requested):
Thanks! Almost there.

* the docstring should spell out the parameter name. It is very confusing to read your docstring, and see the single argument `summarizer` in the constructor, and also as a parameter in your docstring, but none of the other arguments in your docstring.
* I would not call the `summarizer` argument that, but sth like `lag_config`
* the docstring should explain the dictionary format much better. The current docstring format is confusing, it seems to indicate that the user has to provide a dict with keys `index`, `summarizer`, etc.
* the example should spell the params out, like in the original issue.

danbartl commented:
@fkiraly ok tests no longer cancel and ran much faster now 👍 

I tried to address all points, any other comments?

danbartl commented:
After discussion with @fkiraly :

Three options for NAs that were generated during transformation across windows:

Drop NA 
Pad NA
Keep NA

danbartl commented:
Hi @fkiraly , we discussed yesterday that I would try to implement a functionality to drop the first observations that are filled with NAs by WindowSummarizer,  (e.g. we have a rolling mean of 7 days lagged by 14 days - then we would have to drop the first 21 observations in X and y to be consistent). ForecastingPipeline currently only seems to support transforming X, not y (see screens).  There are multiple places we would have to change, not just  in _pipeline.py .

In general, multiple options exist to change that behavior, but we would have to think a bit how to do that - especially the interaction with Imputer as well as repeated calls of WindowSummarizer would have to be discussed.

Therefore, I would like to keep that out of the initial PR and currently only support 
- keep NAs, 
- padding them with the first observation.

What do you think?

![image](https://user-images.githubusercontent.com/19947407/153717353-d40f473c-3ea6-4e38-af6d-61aebff896c9.png)
![image](https://user-images.githubusercontent.com/19947407/153717361-3fb08779-4afb-4f60-a51b-97d5a9f3154e.png)

fkiraly commented:
>Therefore, I would like to keep that out of the initial PR and currently only support

@danbartl, I'm slightly confused, you seem to imply that the padding/filling behaviour is somehow related to dealing with `y`.
Perhaps you can explain if you think I'm not getting it?

To summarize my opinion:
* I don't think this transformer should touch the issue of dealing with `y` at all.
* I think we should implement the three options: pad, leave NA, remove rows in `X` (as returned in `Xt` by `_transform`)
* let's deal with the pipeline later, if we set the default to pad or leave NA, the pipeline should have no problems

danbartl commented:
@fkiraly lets say we have 500 observations, and we have 500 rows for X and 500 rows for y. If we now apply a window of 21 to that dataset, we will generate one column in X that only has 500-21 = 479 observations. The "drop" option would drop the first 21 observations, but since dropping observations in y is currently unsupported, we will have now a mismatch beween 479 rows for X and 500 rows for Y, and the pipeline will correctly run on error. So for a consistent application of the filter option we would need to filter also Y right?

So I implemented now only the options "None" and "bfill" (carry first observation backwards across shifts / windows), lets think about the next steps next friday.

fkiraly reviewed (changes requested):
Excellent!

docstring comments were addressed.

Minor change request that we should make:
* `y_inner_mtype` should be `None`, since `y` is not used internally.
* `X-y-must-have-same-index` should be `False`, since `y` is ignored

A bigger question, not necessarily a change request:
* to me, `_fit` seems only to prepare internal objects, but doesn't set model parameters (ending in underscore). Hence, should this all not be in `_transform`, and `fit-in-transform` set to True? Naively, lagging does not require fitting, so I feel it should be that kind of transformer?

danbartl commented:
@fkiraly  sorry the first two were a regression when I tested filtering Y.

Integrating _fit makes sense, done!
